lang/cs/Org.Apache.REEF.Network/Elastic/Task/Workflow.cs

```diff
@@ -0,0 +1,298 @@
+
+                if (iteratorOperator.MoveNext())
+                {
+                    _position++;
+                    ResetOperatorPositions();
+
+                    return true;
+                }
+                else
+                {
+                    if (_iteratorsPosition.Count > 1)
+                    {
+                        _iteratorsPosition.RemoveAt(0);
+                        _position = _iteratorsPosition[0] - 1;
+                    }
+
+                    return false;
+                }
+            }
+
+            // In case we have one or zero iterators (or we are at the last iterator when multiple iterators exists)
+            if (_position >= _operators.Count || (_iteratorsPosition.Count > 1 && _position == _iteratorsPosition[1]))
+            {
+                if (_iteratorsPosition.Count == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    _position = _iteratorsPosition[0] - 1;
+
+                    return MoveNext();
+                }
+            }
+
+            if (_isRescheduled)
+            {
+                Current.OnTaskRescheduled.Invoke();
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Method used to make the framework aware that an exception as been thrown during execution.
+        /// </summary>
+        /// <param name="e">The rised exception</param>
+        public void Throw(Exception e)
+        {
+            if (_cancellationSource.IsCancelled)
+            {
+                LOGGER.Log(Level.Warning, "Workflow captured an exception while cancellation source was true.", e);
+            }
+            else
+            {
+                LOGGER.Log(Level.Error, "Workflow captured an exception.", e);
+                _failed = true;
+
+                throw new OperatorException(
+                    "Workflow captured an exception", Current.OperatorId, e, Current.FailureInfo);
+            }
+        }
+
+        /// <summary>
+        /// Start the execution of the workflow from the first operator / iterator.
+        /// </summary>
+        public void Reset()
+        {
+            if (_iteratorsPosition.Count > 0)
+            {
+                _position = _iteratorsPosition[0];
+            }
+            else
+            {
+                _position = 0;
+            }
+        }
+
+        /// <summary>
+        /// Get the current elastic operator.
+        /// </summary>
+        public IElasticOperator Current
+        {
+            get
+            {
+                return _position == -1 ? _operators[0] : _operators[_position];
+            }
+        }
+
+        object IEnumerator.Current
+        {
+            get { return Current; }
+        }
+
+        /// <summary>
+        /// Dispose the workflow.
+        /// </summary>
+        public void Dispose()
+        {
+            lock (_lock)
```

Comment:
looks like we use `_lock` in `Dispose()` only. That does not prevent us from accessing (and mutating) `_operators` from other threads. Shouldn't we use the lock wherever we access the `_operators` (and maybe other members)?

Reply:
This lock is because dispose is actually called several times and I want to enforce that is called only once.

Reply:
Good point! Fixed.